### ISSUE TITLE
Bootstrap does not work on macOS mojave

### ISSUE THREAD

**Issue by graeme-winter** [cctbx/cctbx_project#240]
Because something moved - /usr/include now empty, cannot find e.g. zlib. This seems to be by design. I have xcode-select installed cl tools etc.

**Comment by graeme-winter** [comment]
https://silvae86.github.io/sysadmin/mac/osx/mojave/beta/libxml2/2018/07/05/fixing-missing-headers-for-homebrew-in-mac-osx-mojave.html

not a unique problem

**Comment by graeme-winter** [comment]
Testing 

```
sudo installer -pkg /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg -target /
```

Common sense suggests this will help.

**Comment by graeme-winter** [comment]
Next broken thing

```
Installing wxPython...
  log file is /Users/graeme/svn/cctbx/base_tmp/wxPython_install_log
  getting package wxPython-src-3.0.2.0.tar.bz2...
    downloading from http://cci.lbl.gov/cctbx_dependencies : 55.6 MB
    [0%.........20%.........40%.........60%.........80%.........100%]
  installing wxPython-src-3.0.2.0.tar.bz2...
  building wxWidgets with options:
    --prefix="/Users/graeme/svn/cctbx/base"
    --with-opengl
    --enable-unicode
    --without-libjbig
    --without-liblzma
    --with-libjpeg=builtin
    --enable-optimize
    --disable-debugreport
    --with-osx_cocoa
    --with-macosx-version-min=10.7
    --with-mac
    --enable-monolithic
    --disable-mediactrl
Traceback (most recent call last):
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 1649, in <module>
    installer(args=sys.argv, log=sys.stdout)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 247, in __init__
    self.build_dependencies(packages=packages)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 765, in build_dependencies
    getattr(self, 'build_%s'%i)()
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 1538, in build_wxpython
    self.call("make -j %d" % self.nproc, log=pkg_log)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 349, in call
    return call(args, log=log, verbose=self.verbose, **kwargs)
  File "/Users/graeme/svn/cctbx/modules/cctbx_project/libtbx/auto_build/installer_utils.py", line 53, in call
    raise RuntimeError("Call to '%s' failed with exit code %d" % (args, rc))
RuntimeError: Call to 'make -j 8' failed with exit code 2
Process failed with return code 1
```

logs:

```
/Users/graeme/svn/cctbx/base_tmp/wxPython-src-3.0.2.0/bk-deps g++ -mmacosx-version-min=10.7 -c -o wxscintilla_ScintillaBase.o -DNDEBUG -I./src/stc/scintilla/include -I./src/stc/scintilla/lexlib -I./src/stc/scintilla/src -D__WX__ -DSCI_LEXER -DLINK_LEXERS  -D__WXOSX_COCOA__   -dynamic -fPIC -DPIC -D_FILE_OFFSET_BITS=64 -I/Users/graeme/svn/cctbx/base_tmp/wxPython-src-3.0.2.0/lib/wx/include/osx_cocoa-unicode-3.0 -I./include -I/Users/graeme/svn/cctbx/base/include/openssl -I/Users/graeme/svn/cctbx/base/include -I/Users/graeme/svn/cctbx/base/include/freetype2 -stdlib=libc++ -mmacosx-version-min=10.7 -DWX_PRECOMP -O2 -fno-strict-aliasing -fno-common  ./src/stc/scintilla/src/ScintillaBase.cxx
./src/stc/scintilla/src/Editor.cxx:5844:6: error: call to 'abs' is ambiguous
        if (abs(pt1.x - pt2.x) > 3)
            ^~~
/usr/include/stdlib.h:132:6: note: candidate function
int      abs(int) __pure2;
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h:111:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h:113:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
./src/stc/scintilla/src/Editor.cxx:5846:6: error: call to 'abs' is ambiguous
        if (abs(pt1.y - pt2.y) > 3)
            ^~~
/usr/include/stdlib.h:132:6: note: candidate function
int      abs(int) __pure2;
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h:111:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h:113:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
2 errors generated.
/Users/graeme/svn/cctbx/base_tmp/wxPython-src-3.0.2.0/bk-deps g++ -mmacosx-version-min=10.7 -c -o wxscintilla_Selection.o -DNDEBUG -I./src/stc/scintilla/include -I./src/stc/scintilla/lexlib -I./src/stc/scintilla/src -D__WX__ -DSCI_LEXER -DLINK_LEXERS  -D__WXOSX_COCOA__   -dynamic -fPIC -DPIC -D_FILE_OFFSET_BITS=64 -I/Users/graeme/svn/cctbx/base_tmp/wxPython-src-3.0.2.0/lib/wx/include/osx_cocoa-unicode-3.0 -I./include -I/Users/graeme/svn/cctbx/base/include/openssl -I/Users/graeme/svn/cctbx/base/include -I/Users/graeme/svn/cctbx/base/include/freetype2 -stdlib=libc++ -mmacosx-version-min=10.7 -DWX_PRECOMP -O2 -fno-strict-aliasing -fno-common  ./src/stc/scintilla/src/Selection.cxx
make: *** [wxscintilla_Editor.o] Error 1
make: *** Waiting for unfinished jobs....
```

Advice: do not try to run a fresh bootstrap on Mojave...

**Comment by graeme-winter** [comment]
https://github.com/cctbx/cctbx_project/issues/170 would probably solve this

**Comment by graeme-winter** [comment]
@Anthchirp points out on a side channel that this issue may be simpler than it seems:

https://github.com/cctbx/cctbx_project/blame/master/libtbx/auto_build/install_base_packages.py#L1463

**Comment by graeme-winter** [comment]
if (self.flag_is_mac and get_os_version() in ("10.10", "10.11", "10.12", "10.13")) 

14 != 13, or 12, or 11, or 10

**Comment by Anthchirp** [comment]
putative fix deployed with https://github.com/cctbx/cctbx_project/commit/98fdd87a502c412b57dc16f956de623e204af901

**Comment by phyy-nx** [comment]
@graeme-winter does @Anthchirp's fix work for you?  If so, we can close this?
Thanks

**Comment by graeme-winter** [comment]
Not had a chance to test yet, will do tonight

**Comment by graeme-winter** [comment]
Computer says no:

```
Installing wxPython...
  log file is /Users/graeme/svn/cctbx/base_tmp/wxPython_install_log
  getting package wxPython-src-3.0.2.0.tar.bz2...
    downloading from http://cci.lbl.gov/cctbx_dependencies : 55.6 MB
    [0%.........20%.........40%.........60%.........80%.........100%]
  installing wxPython-src-3.0.2.0.tar.bz2...
  patching src/osx/webview_webkit.mm
  patching src/osx/core/bitmap.cpp
  patching src/osx/carbon/dataobj.cpp
  building wxWidgets with options:
    --prefix="/Users/graeme/svn/cctbx/base"
    --with-opengl
    --enable-unicode
    --without-libjbig
    --without-liblzma
    --with-libjpeg=builtin
    --enable-optimize
    --disable-debugreport
    --with-osx_cocoa
    --with-macosx-version-min=10.7
    --with-mac
    --enable-monolithic
    --disable-mediactrl
    CPPFLAGS="-D__ASSERT_MACROS_DEFINE_VERSIONS_WITHOUT_UNDERSCORES=1 -mmacosx-version-min=10.7"
    LDFLAGS="-mmacosx-version-min=10.7"
Traceback (most recent call last):
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 1649, in <module>
    installer(args=sys.argv, log=sys.stdout)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 247, in __init__
    self.build_dependencies(packages=packages)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 765, in build_dependencies
    getattr(self, 'build_%s'%i)()
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 1537, in build_wxpython
    self.call("./configure %s" % " ".join(config_opts), log=pkg_log)
  File "modules/cctbx_project/libtbx/auto_build/install_base_packages.py", line 349, in call
    return call(args, log=log, verbose=self.verbose, **kwargs)
  File "/Users/graeme/svn/cctbx/modules/cctbx_project/libtbx/auto_build/installer_utils.py", line 53, in call
    raise RuntimeError("Call to '%s' failed with exit code %d" % (args, rc))
RuntimeError: Call to './configure --prefix="/Users/graeme/svn/cctbx/base" --with-opengl --enable-unicode --without-libjbig --without-liblzma --with-libjpeg=builtin --enable-optimize --disable-debugreport --with-osx_cocoa --with-macosx-version-min=10.7 --with-mac --enable-monolithic --disable-mediactrl CPPFLAGS="-D__ASSERT_MACROS_DEFINE_VERSIONS_WITHOUT_UNDERSCORES=1 -mmacosx-version-min=10.7" LDFLAGS="-mmacosx-version-min=10.7"' failed with exit code 1
Process failed with return code 1
```

Invetsigating

**Comment by graeme-winter** [comment]
Error in configure:

```
checking for --enable-dirdlg... yes
checking for --enable-msgdlg... yes
checking for --enable-numberdlg... yes
checking for --enable-splash... yes
checking for --enable-textdlg... yes
checking for --enable-tipdlg... yes
checking for --enable-progressdlg... yes
checking for --enable-wizarddlg... yes
checking for --enable-menus... yes
checking for --enable-miniframe... yes
checking for --enable-tooltips... yes
checking for --enable-splines... yes
checking for --enable-mousewheel... yes
checking for --enable-validators... yes
checking for --enable-busyinfo... yes
checking for --enable-hotkey... auto
checking for --enable-joystick... yes
checking for --enable-metafile... auto
checking for --enable-dragimage... yes
checking for --enable-accessibility... no
checking for --enable-uiactionsim... yes
checking for --enable-dctransform... yes
checking for --enable-webviewwebkit... yes
checking for --enable-palette... yes
checking for --enable-image... yes
checking for --enable-gif... yes
checking for --enable-pcx... yes
checking for --enable-tga... yes
checking for --enable-iff... yes
checking for --enable-pnm... yes
checking for --enable-xpm... yes
checking for --enable-ico_cur... yes
checking for --enable-dccache... yes
checking for --enable-ps-in-msw... yes
checking for --enable-ownerdrawn... yes
checking for --enable-uxtheme... yes
checking for --enable-wxdib... yes
checking for --enable-webviewie... yes
checking for --enable-autoidman... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether we are using the Intel C compiler... no
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking whether gcc needs -traditional... no
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking whether we are using the Intel C++ compiler... no
checking whether we are using the Metrowerks C++ compiler... no
checking whether we are using the IBM xlC C++ compiler... no
checking for ar... ar
checking if C compiler (gcc -mmacosx-version-min=10.7) works with SDK/version options... yes
checking if C++ compiler (g++ -mmacosx-version-min=10.7) works with SDK/version options... yes
checking if CoreFoundation/CFBase.h is usable... yes
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for langinfo.h... yes
checking for wchar.h... yes
checking for sys/select.h... yes
checking for cxxabi.h... yes
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking size of short... 2
checking size of void *... 8
checking size of int... 4
checking size of long... 8
checking size of size_t... 8
checking size of long long... 8
checking size of wchar_t... 4
checking for va_copy... yes
checking whether the compiler supports variadic macros... yes
checking for _FILE_OFFSET_BITS value needed for large files... 64
checking if large file support is available... yes
checking for _LARGEFILE_SOURCE value needed for large files... no
checking whether byte ordering is bigendian... no
checking for iostream... no
checking if C++ compiler supports the explicit keyword... yes
checking for std::wstring in <string>... no
checking if std::basic_string<wchar_t> works... no
configure: error: Can't use --enable-std_string without std::wstring or std::basic_string<wchar_t>
```

**Comment by Anthchirp** [comment]
given that there is no schedule for the conda move it seems inappropriate to hold this issue for it, especially since [Mojave will likely to be the most widely used MacOS distribution by the end of the year](http://gs.statcounter.com/macos-version-market-share/desktop/worldwide).

**Comment by graeme-winter** [comment]
Configuring a macOS Mojave system at the moment so we can debug this

**Comment by jhagege** [comment]
+1 any solution found ?

**Comment by graeme-winter** [comment]
@ndevenish if I run the command you copy above it _should_ work yes? I will try this from home.

**Comment by graeme-winter** [comment]
Bodes well:

```
Installing wxPython 4.0.3...
Cross platform GUI toolkit for Python, "Phoenix" version
  log file is /Users/graeme/svn/cctbx-test/base_tmp/wxPython_install_log
  verifying wxPython installation in /Users/graeme/svn/cctbx-test/base/bin/python OK
  package wxpython took 12.5s to install
```

Has moved on to building actual cctbx code, which has to be a good thing; once I have made sure GUI tools in dials work I will update here

**Comment by phyy-nx** [comment]
Hi Graeme, isn't there a copy of the slip viewer in DIALS right now?
Separate from spotfinder_frame which subclasses the slip viewer.  Do you
mean that changes to support wx4 be ported back to the cctbx slip viewer,
and the extra copy in DIALS be removed?

**Comment by graeme-winter** [comment]
@phyy-nx yes, there is a copy in DIALS - which is why the wxPython4 shot above exists. Guess back-porting the changes should not be difficult. We made the copy precisely _because_ we wanted to decouple to allow changes like this and others. What happens now I guess is a discussion topic.

**Comment by ndevenish** [comment]
> Other news - the image viewer should probably be made "dark mode aware"

I believe that there is a load of changes merged into wxWidgets to deal with this - wxPython will pick these changes up in [wxPython 4.1](https://github.com/wxWidgets/Phoenix/issues/1020#issuecomment-430050234)

**Comment by alyubimov** [comment]
Quick note: PyPlot from MatPlotLib 2.0.0 doesn't do well with wxPython 4. If displaying a plot with PyPlot, best use the 'TKAgg' backend. If using the 'WXAgg' backend, PyPlot will try to use some wxPython 4 toobar widgets and crash. Embedded plots seem OK, on the other hand.

**Comment by graeme-winter** [comment]
@alyubimov guessing that this is not an issue in 3.0.2 (Python3) or 2.2.x (Python2)?

**Comment by alyubimov** [comment]
The issue does seem to be resolved (according to MatPlotLib webpage anyway) for 3.0.2. However, the Python2 version of cctbx seems to have 2.0.0 installed, not 2.2.x. At least that's what I get when I do this (with my --builder=dials install) :

$ libtbx.python
Python 2.7.15 (default, Nov 14 2018, 20:28:47) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib
>>> matplotlib.__version__
'2.0.0'

My month-old --builder=xfel install says the same thing.

**Comment by Anthchirp** [comment]
We currently build Matplotlib 2.0.0 on mac and linux, so this is not a suprise.
For DIALS we made sure some time ago that we are Matplotlib 2.2.2 compatible (#191) as someone mentioned there was consideration of whether to update. CCP4 apparently will update Matplotlib next year as well. From our experience only minimal changes were required (see comment in #191 and commit https://github.com/dials/dials/commit/ff05d0fba918828a4ad8de3fbc43f88b1364dcbf).

If this is now more or less required for new MacOS builds we should probably consider updating Matplotlib to 2.2.3 LTS for all platforms.

**Comment by graeme-winter** [comment]
@nksauter I think the changes in the image viewer code were:

https://github.com/dials/dials/commit/763925e668fd46154e61927a10bb84e0d9eeee5b
https://github.com/dials/dials/commit/a0374ab41df87802827500c50b5d733541da3ecd

(pulled from git log dials/command_line/image_viewer.py)

@ndevenish made these changes; possibly I have missed one but I would expect this is the complete list of relevant changes

**Comment by graeme-winter** [comment]
Given that bootstrap _does_ work on Mojave with the addition of the wxPython4 flag, I think all we should do is document this and close the ticket. 

Making the cctbx image viewer work with wxPython4 (perhaps derived from the change sets above) should be a different ticket?

**Comment by graeme-winter** [comment]
Works enough for me, got a build running & issues with wxPython4 for other graphical tools should have their own (more specific) tickets.

**Comment by rjgildea** [comment]
Reopening this ticket as bootstrap still does not work out of the box on Mojave, resulting in wasted hours trying to debug. Can we make wxpython4 the default on mojave, and also documenting the necessary extra steps required to obtain header files in the necessary location?

**Comment by ndevenish** [comment]
See also #298 regarding wxpython4 - there seems no way to get the bootstrap conda to configure it.

**Comment by rjgildea** [comment]
Resolved by d6f0f74 and e9d68f7.
